cpp/src/io/json/json_tree.cu

```diff
@@ -0,0 +1,280 @@
+    node_range_tuple_it,
+    node_ranges_t{tokens.begin(), token_indices.begin(), num_tokens, include_quote_char});
+
+  auto node_range_out_end =
+    thrust::copy_if(rmm::exec_policy(stream),
+                    thrust::make_counting_iterator<size_type>(0),
+                    thrust::make_counting_iterator<size_type>(0) + num_tokens,
+                    node_range_out_it,
+                    [is_node, tokens_gpu = tokens.begin()] __device__(size_type i) -> bool {
+                      PdaTokenT const token = tokens_gpu[i];
+                      return is_node(token);
+                    });
+  CUDF_EXPECTS(node_range_out_end - node_range_out_it == num_nodes, "node range count mismatch");
+
+  // Node parent ids: previous push token_id transform, stable sort, segmented scan with Max,
+  // reorder, copy_if. This one is sort of logical stack. But more generalized.
+  // TODO: make it own function.
+  rmm::device_uvector<size_type> parent_token_ids(num_tokens, stream);
+  rmm::device_uvector<size_type> initial_order(num_tokens, stream);
+  thrust::sequence(rmm::exec_policy(stream), initial_order.begin(), initial_order.end());
+  thrust::tabulate(rmm::exec_policy(stream),
+                   parent_token_ids.begin(),
+                   parent_token_ids.end(),
+                   [does_push, tokens_gpu = tokens.begin()] __device__(auto i) -> size_type {
+                     if (i == 0)
+                       return -1;
+                     else
+                       return does_push(tokens_gpu[i - 1]) ? i - 1 : -1;
+                   });
+  auto out_pid = thrust::make_zip_iterator(parent_token_ids.data(), initial_order.data());
+  // Uses radix sort for builtin types.
+  thrust::stable_sort_by_key(rmm::exec_policy(stream),
+                             token_levels.data(),
+                             token_levels.data() + token_levels.size(),
+                             out_pid);
+  // SegmentedScan Max.
+  thrust::inclusive_scan_by_key(rmm::exec_policy(stream),
+                                token_levels.data(),
+                                token_levels.data() + token_levels.size(),
+                                parent_token_ids.data(),
+                                parent_token_ids.data(),  // size_type{-1},
+                                thrust::equal_to<size_type>{},
+                                thrust::maximum<size_type>{});
+  // FIXME: Avoid sorting again by scatter + extra memory. Tradeoff?
+  thrust::sort_by_key(rmm::exec_policy(stream),
+                      initial_order.data(),
+                      initial_order.data() + initial_order.size(),
+                      parent_token_ids.data());
+  // thrust::scatter(rmm::exec_policy(stream),
+  //                parent_token_ids.begin(),
+  //                parent_token_ids.end(),
+  //                initial_order.data(),
+  //                parent_token_ids.begin()); //same location not allowed in scatter
```

Comment:
Dead code?
```suggestion
```

Reply:
Yes. already FIXME is added for this part. it's a TODO for later.